File: lab/control/themes.go
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Hashem Aldhaheri

package main

import (
	"encoding/json"
	"io/fs"
	"log"
	"path"
	"sort"
)

// ---------------------------------------------------------------------------
// HUD themes
//
// The board has always read its colours off the page rather than carrying a
// palette of its own — readTokens() in hud/scene.js pulls nine custom
// properties and paints from them. That was done so the light/dark button
// could move the board with it, and it is the whole mechanism a theme needs:
// redefine the nine, tell the board they moved.
//
// So a theme is a folder under ui/hud/themes/ with two files in it. theme.css
// redefines whichever of the nine it cares about, scoped to
// :root[data-hud-theme="<id>"] so it is inert until selected. theme.json says
// what it is called and which ground it was drawn for.
//
// This file is the half that finds them. It walks the embedded UI at startup
// so the page does not have to be told what exists — adding a theme is a
// folder and a rebuild, not an edit to a list in three places. Which is the
// same reason panels.go exists: the server holds the spec, the UI renders
// whatever it is handed.
//
// A theme that does not parse, or whose id disagrees with its folder, is
// skipped with a line at the console rather than taken as a reason to refuse
// to start. That is the choice setpieces.go's missing() already makes and for
// the same reason: a broken extra must not take the lab with it, and a silent
// one is how `rotate-key` went a release with nothing drawn.
// ---------------------------------------------------------------------------

// Theme is one entry in the store, as /api/themes serves it.
type Theme struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	Note string `json:"note"`
	// Ground is the page background the theme was drawn against: "light",
	// "dark", or "any" for one that works on either. It is advisory — selecting
	// a theme that names a ground moves the page to it, and the light/dark
	// button still wins afterwards.
	Ground string `json:"ground"`
	Author string `json:"author"`
}

// themesDir is where the store lives inside the UI filesystem. It is a
// constant rather than a flag because the UI is //go:embed-ed: nothing can put
// a theme anywhere else without rebuilding the binary, which is the point.
const themesDir = "hud/themes"

// baseTheme is the one that must exist. It overrides nothing — it is the
// guide's own palette, which is what the board looked like before there was a
// store — so it is also the fallback when a saved id no longer resolves.
const baseTheme = "chapter"

var validGrounds = map[string]bool{"light": true, "dark": true, "any": true}

// loadThemes walks fsys for ui/hud/themes/*/theme.json and returns what it
// finds, base theme first and the rest alphabetically by name.
//
// fsys is the UI sub-filesystem, so paths here start at hud/.
func loadThemes(fsys fs.FS) []Theme {
	entries, err := fs.ReadDir(fsys, themesDir)
	if err != nil {
		log.Printf("hud themes: %s is not readable (%v), so the store is empty", themesDir, err)
		return nil
	}

	var out []Theme
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		t, ok := readTheme(fsys, e.Name())
		if !ok {
			continue
		}
		out = append(out, t)
	}

	sort.Slice(out, func(i, j int) bool {
		if (out[i].ID == baseTheme) != (out[j].ID == baseTheme) {
			return out[i].ID == baseTheme
		}
		return out[i].Name < out[j].Name
	})
	return out
}

// readTheme reads one folder, and says why it skipped it when it does.
//
// The id has to match the folder name because the folder name is what the
// page builds the stylesheet URL from. A theme whose json claimed a different
// id would load one file and select another, and the failure would be a board
// that keeps the previous palette with no error anywhere.
func readTheme(fsys fs.FS, dir string) (Theme, bool) {
	b, err := fs.ReadFile(fsys, path.Join(themesDir, dir, "theme.json"))
	if err != nil {
		log.Printf("hud theme %q: no theme.json, skipped", dir)
		return Theme{}, false
	}
	var t Theme
	if err := json.Unmarshal(b, &t); err != nil {
		log.Printf("hud theme %q: theme.json does not parse (%v), skipped", dir, err)
		return Theme{}, false
	}
	if t.ID != dir {
		log.Printf("hud theme %q: theme.json says id %q, skipped", dir, t.ID)
		return Theme{}, false
	}
	if t.Name == "" {
		log.Printf("hud theme %q: no name, skipped", dir)
		return Theme{}, false
	}
	if _, err := fs.Stat(fsys, path.Join(themesDir, dir, "theme.css")); err != nil {
		log.Printf("hud theme %q: no theme.css, skipped", dir)
		return Theme{}, false
	}
	if !validGrounds[t.Ground] {
		// Not a reason to drop the theme: "any" is the honest reading of a
		// theme that did not say, and it is what the base theme is.
		log.Printf("hud theme %q: ground %q is not light, dark or any — read as any", dir, t.Ground)
		t.Ground = "any"
	}
	return t, true
}

File: lab/control/ui/hud/themes.js
/* SPDX-License-Identifier: GPL-3.0-or-later
   Copyright (C) 2026 Hashem Aldhaheri */

/* ============================================================
   The HUD theme store.

   A theme is a folder under hud/themes/ holding theme.css and theme.json.
   themes.go walks the embedded UI and serves what it found at /api/themes,
   so nothing here carries a list: the server knows, and this asks.

   Selecting a theme does three things and no more.

     1. Make sure the theme's stylesheet is in the document. Every theme's
        rules are scoped to :root[data-hud-theme="<id>"], so a sheet that has
        been loaded and then deselected is inert rather than removed. Loading
        is one-way, which keeps switching back instant and keeps this file
        free of the question of when a <link> is safe to drop.

     2. Set data-hud-theme on the root element, and — if the theme named a
        ground — move data-theme to it. Advisory, as themes.go says: the
        light/dark button still wins afterwards, because a reader who presses
        it has said something more recent than the theme's manifest did.

     3. Tell the board its colours moved. scene.js reads the nine --board-*
        properties once and paints from what it read, so nothing changes on
        screen until refreshTokens() is called. This is the same call the
        light/dark button already makes.

   Every entry point survives having no board, no server and no localStorage,
   because the page it belongs to does. A control surface that will not draw
   because a stylesheet 404'd is a worse failure than an unthemed board.
   ============================================================ */

const BASE = "chapter";           /* matches baseTheme in themes.go */
const STORE_KEY = "lab-hud-theme";

/* Which stylesheets have been put in the document, by id. */
const loaded = new Set();

let catalog = [];
let active = BASE;

/* ---------- what exists ---------------------------------------- */

/* The store, as the server found it. Empty until load() has run, which is
   the honest answer before then rather than a guess. */
export function list() { return catalog.slice(); }

/* The id currently selected. */
export function current() { return active; }

/* One entry by id, or null. */
export function get(id) {
  for (const t of catalog) if (t.id === id) return t;
  return null;
}

/* Ask the server what it embedded. A failure here is not fatal: the base
   theme overrides nothing, so a page with no catalog still draws the board
   exactly as it did before this file existed. */
export async function load() {
  try {
    const res = await fetch("/api/themes", { headers: { Accept: "application/json" } });
    if (!res.ok) throw new Error("HTTP " + res.status);
    const body = await res.json();
    catalog = Array.isArray(body) ? body : [];
  } catch (e) {
    console.warn("hud themes: /api/themes did not answer (" + e.message + "), staying on " + BASE);
    catalog = [];
  }
  return list();
}

/* Hand the store over rather than asking for it.
 *
 * The player page has no server to ask. A recording carries the store in its
 * opening hello — see Page in lab/control/feed.go — precisely so that a
 * captured attack can be watched with the cards the theme docked, and this is
 * where that list goes in. Everything after it is identical: the stylesheets
 * are still fetched by relative path, because the player is served from the
 * same directory the live page is.
 *
 * Returns what was accepted, so a caller can see whether the recording
 * carried anything at all. */
export function seed(store) {
  catalog = Array.isArray(store) ? store.slice() : [];
  return list();
}

/* ---------- selecting one -------------------------------------- */

/* Put a theme's stylesheet in the document, and resolve when the browser has
   parsed it. The wait matters: refreshTokens() reads computed values, and a
   board told to repaint before its palette landed paints the old one and has
   no reason to try again.

   A sheet that fails to load resolves too. The rules simply never apply, the
   board keeps the palette it had, and the console says which file was
   missing — which is a readable board and a legible complaint, rather than a
   promise that never settles. */
function ensureSheet(id) {
  if (loaded.has(id)) return Promise.resolve();
  return new Promise((resolve) => {
    const link = document.createElement("link");
    link.rel = "stylesheet";
    link.href = "hud/themes/" + encodeURIComponent(id) + "/theme.css";
    link.dataset.hudTheme = id;
    link.addEventListener("load", () => { loaded.add(id); resolve(); });
    link.addEventListener("error", () => {
      console.warn("hud theme " + id + ": theme.css did not load, the board keeps its colours");
      /* Marked loaded on purpose: retrying on every switch would add one dead
         <link> per press for as long as the file is missing. */
      loaded.add(id);
      resolve();
    });
    document.head.appendChild(link);
  });
}

/* Select a theme. `board` may be null — the page runs without one whenever
   WebGL is unavailable, and the attribute is still worth setting so anything
   else styled off it agrees with the picker. */
export async function apply(id, board) {
  const theme = get(id) || get(BASE);
  const next = theme ? theme.id : BASE;

  await ensureSheet(next);
  document.documentElement.setAttribute("data-hud-theme", next);

  if (theme && (theme.ground === "light" || theme.ground === "dark")) {
    document.documentElement.setAttribute("data-theme", theme.ground);
    try { localStorage.setItem("lab-theme", theme.ground); } catch (e) { /* private mode */ }
  }

  active = next;
  try { localStorage.setItem(STORE_KEY, next); } catch (e) { /* private mode */ }

  if (board) board.refreshTokens();
  return next;
}

/* ---------- starting up ---------------------------------------- */

/* Read the catalog, then select whatever was chosen last — falling back to
   the base theme when the saved id names a theme that is no longer embedded,
   which is what happens the first time somebody removes a folder.

   Returns the id that ended up applied, so a caller can set its picker from
   the answer rather than from what it asked for.

   `store` is for the player page: hand it the list a recording carried and
   nothing is fetched. The rest of this function is the part worth sharing —
   remembering which theme the reader picked, and noticing when the saved id
   names a theme that is not in front of it. Both pages want that and neither
   should own a copy of it. */
export async function init(board, store) {
  if (store) seed(store);
  else await load();

  let saved = null;
  try { saved = localStorage.getItem(STORE_KEY); } catch (e) { /* private mode */ }

  if (saved && !get(saved)) {
    console.warn("hud themes: " + saved + " is no longer embedded, falling back to " + BASE);
    saved = null;
  }
  return apply(saved || BASE, board);
}
